sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -301,12 +317,170 @@ case class GpuCaseWhen(
+              }
+
+            case None =>
+              // if there is no ELSE condition then we return NULL for any rows not matched by
+              // previous branches
+              withResource(GpuScalar.from(null, dataType)) { nullScalar =>
+                if (isAllFalse(cumulativePred.get)) {
+                  GpuColumnVector.from(nullScalar, elsePredNoNulls.getRowCount.toInt, dataType)
+                } else {
+                  GpuColumnVector.from(
+                    elsePredNoNulls.ifElse(nullScalar, currentValue.get.getBase),
+                      dataType)
+                }
+              }
+          }
+        }
+      }
+    } finally {
+      currentValue.foreach(_.safeClose())
+      cumulativePred.foreach(_.safeClose())
+    }
+  }
+
+  /**
+   * Filter the batch to just the rows where the WHEN condition is true and
+   * then evaluate the THEN expression.
+   */
+  private def filterEvaluateWhenThen(
+      colTypes: Array[DataType],
+      tbl: Table,
+      whenBool: ColumnVector,
+      thenExpr: Expression): ColumnVector = {
+    val filteredBatch = filterBatch(tbl, whenBool, colTypes)
+    val thenValues = withResource(filteredBatch) { trueBatch =>
+      GpuExpressionsUtils.columnarEvalToColumn(thenExpr, trueBatch)
+    }
+    withResource(thenValues) { _ =>
+      gather(whenBool, thenValues)
+    }
+  }
+
+  /**
+   * Calculate the cumulative predicate so far using the logical expression
+   * `prevPredicate OR thisPredicate`.
+   */
+  private def calcCumulativePredicate(
+      cumulativePred: Option[GpuColumnVector],
+      whenBool: GpuColumnVector,
+      firstTrueWhen: GpuColumnVector): GpuColumnVector = {
+    cumulativePred match {
+      case Some(prev) =>
+        withResource(prev) { _ =>
+          GpuColumnVector.from(whenBool.getBase.or(prev.getBase), DataTypes.BooleanType)
+        }
+      case _ =>
+        firstTrueWhen.incRefCount()
+    }
+  }
+
+  /**
+   * Calculate the current values by merging the THEN values for this branch (where the WHEN
+   * predicate was true) with the previous values.
+   */
+  private def calcCurrentValue(
+      prevValue: Option[GpuColumnVector],
+      whenBool: GpuColumnVector,
+      thenValues: ColumnVector): GpuColumnVector = {
+    prevValue match {
+      case Some(v) =>
+        withResource(v) { _ =>
+          GpuColumnVector.from(whenBool.getBase.ifElse(thenValues, v.getBase), dataType)
+        }
+      case _ =>
+        GpuColumnVector.from(thenValues.incRefCount(), dataType)
+    }
+  }
+
+  /**
+   * Determine for each row whether this is the first WHEN predicate so far to evaluate to true
+   */
+  private def isFirstTrueWhen(
+      cumulativePred: Option[GpuColumnVector],
+      whenBool: GpuColumnVector): GpuColumnVector = {
+    cumulativePred match {
+      case Some(prev) =>
+        withResource(prev.getBase.not()) { notPreviouslyTrue =>
+          GpuColumnVector.from(whenBool.getBase.and(notPreviouslyTrue), DataTypes.BooleanType)
```

Comment:
Similar concern with nulls here.  If `prev` could contain nulls then `not()` will still be null and then `and` will result in a null which is undesired if `whenBool` contains a `true` in the corresponding row.